### ISSUE TITLE
CDC data inconsistencies with schema changes

### ISSUE THREAD

**Issue by bmaisonn** [apache/iceberg#7822]
### Apache Iceberg version

1.2.1 (latest release)

### Query engine

Spark

### Please describe the bug 🐞

We're testing the CDC feature released in version 1.2.1 (also tested with 1.3.0) and we noticed a strange behavior:

We performed the following actions

| snapshot | change type                                 |
|----------|---------------------------------------------|
| s0       | initial insert of record A                  |
| s1       | update of record A on col1                  |
| s2       | ADD column col2 + update record A on col2   |
| s3       | DROP column col2 + update record A on col1  |
| s4       | ADD column 2 + update record A on col2  |

Before we drop the col 2 we see col2 in cdc generated between  S2 and (S1 or S0).
After we drop col2, col2 is not visible anymore in cdc whatever the snapshots given. For instance cdc between S2 and S1 is empty while it was not empty before we drop col2.
If we create col2 again with new data (i.e s4):
* CDC between S2 and S1 col2 is visible but the dataframe is empty
* CDC between S4 and S1 col2 is visible and only the value added in S4 is visible

In this example we would expect to see the changes for col2 between S2 and S1 even after we drop col2 as the schema modification after ? 

To reproduce:

* create a table with few columns and insert records
* add a new column and perform some updates
* drop that last column
* check cdc

**Comment by bmaisonn** [comment]
Hello team, any thought on that one ? We're wondering if our assumption is correct as this specific case isn't described in the documentation. If that's the expected behavior we'll try to find a workaround on our end to avoid loosing logs after the schema is modified.
Thx for your help

**Comment by nastra** [comment]
@stevenzwu could you take a look please?

**Comment by RussellSpitzer** [comment]
Could we get a tiny repo ? Seems like this should just be a few lines of code. Then we could talk about a specific example rather than in abstract

**Comment by flyrain** [comment]
Hi @bmaisonn, it seems a bug to me. I will take a look. Would you like to add a unit test for this bug in `TestCreateChangelogViewProcedure`?

**Comment by bmaisonn** [comment]
@flyrain @RussellSpitzer here is a notebook with a concrete example of what we found out https://github.com/bmaisonn/iceberg_tests/blob/main/cdc.md
We'd be happy contributing but i'm not really sure what could be the good behavior here when the schema is altered. In our use case we'd like to see the changes as they were applied when the snapshots were generated and the corresponding schema but that might be difficult to represent clearly in a table when the changelog spans multiple snapshots. 
Note that our workaround for now will be to ensure that we export the changelog before any schema modification

**Comment by flyrain** [comment]
The cdc procedure is based on the changelog table. The changelog table reader reads rows with the current schema even schemas may be different across multiple snapshots. In that case, it either throws extra columns or return null for missing columns. I think the simplest way now is to check if schemas of all snapshots in a given range are the same, and throw if not.  We can think about a better way to handle different schemas, but I feel it is harder to handle different schema within one procedure call. Users may have to handle that situation by themselves. cc @aokolnychyi

**Comment by aokolnychyi** [comment]
Let me catch up on this issue.

**Comment by aokolnychyi** [comment]
I feel like it is important to note that Iceberg tracks columns by ID and the latest assigned column ID is being tracked in the metadata. 

```
spark.sql("ALTER TABLE s3_catalog.cdc.test DROP COLUMN name")
spark.sql('ALTER TABLE s3_catalog.cdc.test ADD COLUMN name string')
```

The snippet above does NOT restore the original column. Instead, it adds a brand-new column with a fresh ID that happens to have the same name as one of the columns that were dropped earlier. That's why the CDC procedure, which acts according to the current schema, outputs nulls for `name` in old columns. From the Iceberg perspective, original files from the first snapshot did not have `name` as the column ID was different.

This behavior is intentional to avoid surprises when a column is dropped and a new one added a few years after and suddenly old data starts to appear in the query output. It is possible to restore an old column but it is not done via ALTER TABLE ADD/DROP COLUMN.

**Comment by aokolnychyi** [comment]
BTW, it would be possible to take into account the old schema as well.
What is the use case are you looking for, @bmaisonn?

**Comment by bmaisonn** [comment]
@aokolnychyi the behavior mentioned above makes sense to me. In this example we didn't necessarily expect that the two columns with the same name to line up in the changelog. We did that when we noticed that the column disappeared from the changelog after we dropped it to try to figure out if it would come back and how.

The use case here is that we want to keep track of the changes made by the users on those tables for quality purpose. Basically we want to be able to easily answer the question "What happened to that datapoint during the last week" and see the modifications made for that period of time. In that context we'd like to see the records as of that modification.

In that case i think it would be acceptable to see two columns with the same name in the changelog or to return the column ids and let us resolve them afterward. Still it would be difficult to make the difference between "this column was there with null values" vs " this column wasn't there at that time" but we could display the schema changes separately.
